app/actions/service_instance_delete_user_provided.rb

```diff
@@ -0,0 +1,29 @@
+module VCAP::CloudController
+  class ServiceInstanceDeleteUserProvided
+    class AssociationNotEmptyError < StandardError; end
+
+    def initialize(event_repo)
+      @service_event_repository = event_repo
+    end
+
+    def delete(service_instance)
+      association_not_empty! if service_instance.has_bindings? || service_instance.has_keys?
+
+      service_instance.db.transaction do
+        service_instance.lock!
+        service_instance.destroy
```

Comment:
I think a `delete()` would be safer, but it's not worth changing.

Reply:
`destroy` runs any hooks (like `before_destroy` and `after_destroy`) that may have been defined in the model -- it also performs validations and etc. `delete` runs the SQL DELETE and that's it. In general (looking through other delete actions) we use `destroy` instead of `delete`